crates/bevy_transform/src/systems.rs

```diff
@@ -46,21 +46,29 @@ pub fn propagate_transforms(
             changed |= children_changed;
 
             for child in children.iter() {
-                propagate_recursive(
-                    &global_transform,
-                    &transform_query,
-                    &parent_query,
-                    &children_query,
-                    entity,
-                    *child,
-                    changed,
-                );
+                // SAFETY: Assuming the hierarchy is consistent, we can be sure that each `child` entity
+                // will be unique across each invocation of this loop body.
+                // Since this is the only place where `transform_query` gets used, `child` will not be borrowed anywhere else.
+                // Also, since `child` only has one ancestor, none of its descendants will be borrowed by other invocations of the loop.
+                unsafe {
+                    propagate_recursive(
+                        &global_transform,
+                        &transform_query,
+                        &parent_query,
+                        &children_query,
+                        entity,
+                        *child,
+                        changed,
+                    );
+                }
             }
         },
     );
 }
 
-fn propagate_recursive(
+/// # Safety
+/// `unsafe_transform_query` must not be borrowed for `entity`, nor any of its descendants.
```

Comment:
```suggestion
/// `unsafe_transform_query` must not fetch aliased mutable references for `entity`, nor any of its descendants.
```

A similar rewording of the other comments should be applied too.

Reply:
I don't think this is any clearer. It just uses more jargon to say the same thing, and it isn't more precise (read-only borrows are also unsound, and aliasing is implied).

I could change it to say 'fetched' instead of 'borrowed'. Do you think that's more clear?

Reply:
Yes. The borrow suggests the Query object itself cannot be borrowed, which it already is at the callsites. We're explicitly talking about fetching the corresponding components for the entity and its subtree, and the language here should be unambiguous.